### PR TITLE
Feature/custom map

### PR DESCRIPTION
**Pull request by Kurtzmusch**

### CODE REVIEW

**Comment by ScorpioneOrzion** [comment]
Error is the same as before

File: server/services/gameCreate.js [hunk 1]
```diff
@@ -59,12 +59,18 @@ module.exports = class GameCreateService {
             settings
         });
 
-        // Calculate how many stars we need.
+        let isCustomGalaxy = game.settings.galaxy.galaxyType === 'custom';
```
Comment (mike-eason) [inline server/services/gameCreate.js:62]: Would be better as a function in `gameService`. i.e `this.gameService.isCustomGalaxy(game)`

File: server/services/map.js [hunk 2]
```diff
@@ -40,32 +41,52 @@ module.exports = class MapService {
...
                 break;
+            case 'custom':
+                starLocations = this.customMapService.generateLocations(game, settings, starCount, playerLimit)
+                break;
             default:
                 throw new ValidationError(`Galaxy type ${game.settings.galaxy.galaxyType} is not supported or has been disabled.`);
         }
 
-        // Iterate over all star locations
+        let isCustomGalaxy = game.settings.galaxy.galaxyType === 'custom';
```
Comment (mike-eason) [inline server/services/map.js:51]: Use the `gameService`? Or pass it into the function?

File: server/services/maps/custom.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+        const locations = [];
+        const playerIndexes = [];
+        //const nameList = new Set()
+        const homeStars = []
+
+        for (const star of json.stars) {
+            //if (nameList.has(star.name)) continue
+
+            if (!this._checkStarProperty(star?.location, 'x', 'number')) continue
```
Comment (mike-eason) [inline server/services/maps/custom.js:24]: Do these need to be if statements? It'll throw an error if the property check fails.

File: server/services/maps/custom.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+            if (!this._checkStarProperty(star?.infrastructure, 'industry', 'number')) continue
+            if (!this._checkStarProperty(star?.infrastructure, 'science', 'number')) continue
+            //if (!this._checkStarProperty(star, 'ships', 'number')) continue
+            if (!this._checkStarProperty(star, 'playerIndex', 'number')) continue
+            if (!this._checkStarProperty(star, 'homeStar', 'boolean')) continue
+            if (!this._checkStarProperty(star, 'specialistId', 'number')) continue
+
+            if (star.naturalResources < 0 || star.naturalResources > 100)
+                throw new ValidationError('Illigal starting amount of resources, range needs to be between 10 and 100 inclusive')
+            if (star.playerIndex >= (8*4)) //colours*shapes
```
Comment (mike-eason) [inline server/services/maps/custom.js:38]: `8*4` - Can we avoid magic numbers? Count the actual colours and shapes we have from the `colours.json` file?

File: server/services/maps/custom.js [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+              star.linkedLocations = []
+              if (star.playerIndex >=0) { playerIndexes.push(star.playerIndex) }
+            }
+            //nameList.add(star.name)
+
+            locations.push(star)
+        }
+        /*
+        if (!ownedList.every(array => array.length === starCount / playerCount))
+            throw new ValidationError('Not enough stars per player.')
```
Comment (mike-eason) [inline server/services/maps/custom.js:52]: Do we need these?

File: server/services/maps/custom.js [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        }
+
+        let commonStarAmount = homeStars[0].linkedLocations.length
+        for (let homeStar of homeStars) {
+          if(homeStar.linkedLocations.length !== commonStarAmount)
+            throw new ValidationError('unfair start - players dont have the same amount of stars')
+        }
+
+        for(let location of locations) {
+          if(location.specialistId === -1) { location.specialistId = null; }
```
Comment (mike-eason) [inline server/services/maps/custom.js:77]: What is this for?

File: server/services/player.js [hunk 7]
```diff
@@ -107,20 +107,32 @@ module.exports = class PlayerService extends EventEmitter {
...
 
-        for(let i = 0; i < game.settings.general.playerLimit; i++) {
-            // Get a random colour to assign to the player.
-            if (!colours.length) {
-                colours = require('../config/game/colours').slice();
-                shapeIndex++;
-            }
+        if(game.galaxy.playerIndexes.length) {
+          for(let index of game.galaxy.playerIndexes) {
+            let colour = colours.splice(index%8, 1)[0];
```
Comment (mike-eason) [inline server/services/player.js:112]: Avoid magic numbers here, store the actual count of colours somewhere above.

File: server/services/player.js [hunk 8]
```diff
@@ -107,20 +107,32 @@ module.exports = class PlayerService extends EventEmitter {
...
+            //let colour = colours[index%8][0] // 0 for alias
+            let shape = shapes[Math.floor(index/8)]
```
Comment (mike-eason) [inline server/services/player.js:114]: 🧙

File: server/services/star.js [hunk 9]
```diff
@@ -51,14 +64,14 @@ module.exports = class StarService extends EventEmitter {
...
 
         this.resetIgnoreBulkUpgradeStatuses(homeStar);
         
         // ONLY the home star gets the starting infrastructure.
-        homeStar.infrastructure.economy = gameSettings.player.startingInfrastructure.economy;
-        homeStar.infrastructure.industry = gameSettings.player.startingInfrastructure.industry;
-        homeStar.infrastructure.science = gameSettings.player.startingInfrastructure.science;
+        homeStar.infrastructure.economy = homeStar.infrastructure.economy ?? gameSettings.player.startingInfrastructure.economy;
+        homeStar.infrastructure.industry = homeStar.infrastructure.industry ?? gameSettings.player.startingInfrastructure.industry;
+        homeStar.infrastructure.science = homeStar.infrastructure.science ?? gameSettings.player.startingInfrastructure.science;
```
Comment (mike-eason) [inline server/services/star.js:74]: We need to be mindful that there might be other game settings that could interfere with custom galaxies, have you checked other settings besides the starting infrastructure?

**Review by mike-eason — CHANGES_REQUESTED** [review]
Needs conflicts resolved!

File: server/services/maps/custom.js (hunk 4 above)
Comment (ScorpioneOrzion) [reply server/services/maps/custom.js:38]: Possible but would the not be better to add a `shape.json` as well?
And write down there the steps as an path2d array?

File: server/services/player.js (hunk 8 above)
Comment (ScorpioneOrzion) [reply server/services/player.js:114]: same as above lol

File: server/services/maps/custom.js (hunk 6 above)
Comment (ScorpioneOrzion) [reply server/services/maps/custom.js:77]: -1 means no specialist

File: server/services/gameCreate.js (hunk 1 above)
Comment (Kurtzmusch) [reply server/services/gameCreate.js:62]: looks like you already changed this?

File: server/services/maps/custom.js (hunk 3 above)
Comment (Kurtzmusch) [reply server/services/maps/custom.js:24]: removed those if's

**Pull request merged**